### PR TITLE
Integrating new external messaging email

### PR DESCRIPTION
**Pull request by stkem**
Not entirely complete, but getting there. This also includes some future utils (in PimpedFuture.scala)

cc @martinraison @andrewconner

### CODE REVIEW

**Comment by stkem** [comment]
@martinraison or @andrewconner can I get a quick review? (Even though this pull request isn't quite up to new old standards yet)
Would like to merge tonight or tomorrow morning so QA can get started.

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala [hunk 1]
```diff
@@ -40,9 +41,76 @@ object PimpMyFuture {
...
   }
 
+
 }
 
+object FutureHelpers {
+  import PimpMyFuture._
+
+  def map[A,B](in: Map[A,Future[B]]): Future[Map[A,B]] = {
+    val seq : Seq[Future[(A, B)]]= in.toSeq.map{ case (key, fut) =>
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala]: `toSeq` not needed.

```
scala> Map(1->2, 3->4).map { case (a,b) => println(s"$a $b") }
1 2
3 4
```

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala [hunk 2]
```diff
@@ -40,9 +41,76 @@ object PimpMyFuture {
...
+      fut.onComplete{ t =>
+        val withKey : Try[(A,B)] = t.map((key, _))
+        p.complete(withKey)
+      }
+      p.future
+    }
+    Future.sequence(seq).map(_.toMap)
+  }
+
+  def combine[T1, T2](f1: Future[T1], f2: Future[T2]): Future[(T1,T2)] = {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala]: Better than `Future.sequence`?

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala]: erm, for comprehension. Offline, we decided we didn't need it.

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala]: Not aware the context, but curious: 
How do we obtain an object like `Map[A,Future[B]]` in the first place? If we have a `Seq[A]`, this can trigger a `Seq[Future[B]]` or Future[Seq[B]], and we can zip them later? 
If that is the case, we need some helper takes two paramters: `Seq[A]` and a function `A -> Future[B]`  ?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 3]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
+
+      val pageName = thread.nUrl.flatMap( url => URI.parse(url).toOption.flatMap( uri => uri.host.map(_.name)) ).get
 
-      val title = thread.pageTitle.getOrElse(url)
+      val messages = basicMessageCommander.getThreadMessages(thread)
 
-      nuts.foreach{ nut =>
-        if ((exceptAddress.isEmpty || exceptAddress.get != nut.participant.identifier) && !nut.muted) {
-          val mute = "/eliza/email/mute/" + nut.publicId.get
+      nuts.filterNot(_.muted).foreach{ nut =>
```
Comment (yingjieMiao) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:128]: a better name, rather than `nuts`?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 4]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
+
+    val uriSummaryFuture : Future[URISummary] = shoebox.getUriSummary(URISummaryRequest(
+      url = thread.nUrl.get,
+      imageType = ImageType.ANY,
+      minSize = ImageSize(183, 96),
+      withDescription = true,
+      waiting = true,
+      silent = false))
+
+    FutureHelpers.combine(allUsersFuture, allUserImageUrlsFuture, uriSummaryFuture).map{ case (allUsers, allUserImageUrls, uriSummary) =>
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala]: As we mentioned, this is a for comprehension.

Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala]: May be useful for this body to be a separate function. Easier to test with non-future values.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 5]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
+          val threadInfo = ThreadEmailInfo(
+            pageUrl = thread.nUrl.get,
+            pageName = pageName,
+            pageTitle = uriSummary.title.getOrElse(thread.nUrl.get),
+            heroImageUrl = uriSummary.imageUrl.getOrElse("#"),
+            pageDescription = uriSummary.description.getOrElse(""),
+            participants = participants.toSeq,
+            conversationStarter = starterUser.firstName + " " + starterUser.lastName,
+            unsubUrl = unsubUrl,
+            muteUrl = "https://kifi.com/extmsg/email/mute?publicId=" + nut.publicId.get //Todo (Stephen): remove hard coded url with public id
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala]: `https://www.kifi.com` (otherwise, it's a redirect)

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 6]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
-          val mute = "/eliza/email/mute/" + nut.publicId.get
+      nuts.filterNot(_.muted).foreach{ nut =>
+        require(nut.participant.kind == NonUserKinds.email)
+
+        shoebox.getUnsubscribeUrlForEmail(nut.participant.identifier).map{ unsubUrl =>
+
+          val threadInfo = ThreadEmailInfo(
+            pageUrl = thread.nUrl.get,
+            pageName = pageName,
+            pageTitle = uriSummary.title.getOrElse(thread.nUrl.get),
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:136]: title is the full url?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 7]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
+            heroImageUrl = uriSummary.imageUrl.getOrElse("#"),
+            pageDescription = uriSummary.description.getOrElse(""),
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:138]: `heroImageUrl` and `pageDescription` should probably be `Option`. Default values is a template concern (and in our case, we use different templates)

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 8]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
 
+          val threadItems = relevantMessages.filterNot(_.from.isSystem).map{ message =>
+            val CleanedMessage(text, lookHereTexts, lookHereImageUrls) = parseMessage(message.messageText)
+            message.from match {
+              case MessageSender.User(id) => ExtendedThreadItem(allUsers(id).firstName, allUsers(id).lastName, allUserImageUrls(id), text, lookHereTexts, lookHereImageUrls)
+              case MessageSender.NonUser(nup) => {
+                val (first, last) = nup.toNameTuple
+                ExtendedThreadItem(first, last, "https://www.kifi.com/assets/img/ghost.200.png", text, lookHereTexts, Seq.empty)
+              }
+              case _ => throw new Exception("Impossible")
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:159]: Why?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala [hunk 9]
```diff
@@ -68,51 +73,114 @@ class NotificationCommander @Inject() (
...
+            }
+          }.reverse
+
+          val body = views.html.nonUserDigestEmail(threadInfo, threadItems).body
+
+          val magicAddress = "discuss+" + nut.publicId.get + "@kifi.com"
           shoebox.sendMail(ElectronicMail (
             from = EmailAddresses.NOTIFICATIONS,
-            fromName = Some("Kifi"),
+            fromName = Some("Kifi Discussions"),
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:168]: Do we want to list the authors? "Effi via Kifi"

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/EmailOptOutController.scala [hunk 10]
```diff
@@ -64,6 +64,10 @@ class EmailOptOutController @Inject() (
     }.getOrElse(BadRequest)
   }
 
+  def getUnsubscribeUrlForEmail(email: String) = Action { implicit request =>
+    Ok(s"https://www.kifi.com${com.keepit.controllers.website.routes.EmailOptOutController.optOut(commander.generateOptOutToken(GenericEmailAddress(email)))}")
```
Comment (martinraison) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/EmailOptOutController.scala:68]: we should probably have a more robust way of managing urls to avoid hardcoding them. Probably fine for now, though

**Comment by martinraison** [comment]
lgtm! (as long as notifyEmailUsers gets broken down a bit at some point)

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThread.scala [hunk 11]
```diff
@@ -19,6 +19,9 @@ sealed trait NonUserParticipant {
   val kind: NonUserKind
 
   override def toString() = identifier.toString
+
+  //first, last
+  def toNameTuple: (String, String)
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThread.scala]: Why? Always presentation? This means a) limited to two named tokens, b) forces having a default string in the second if it only has first, c) results in code like `if (last.length>0) first + " " + last else first`

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThreadRepo.scala [hunk 12]
```diff
@@ -122,4 +125,13 @@ class NonUserThreadRepoImpl @Inject() (
     }.toOption
   }
 
+  def setLastNotifiedAndIncCount(nut: Id[NonUserThread], dt: DateTime)(implicit session: RWSession): Unit = {
+    (for (row <- rows if row.id===nut) yield row.lastNotifiedAt).update(dt)
+    sqlu"""UPDATE non_user_thread
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThreadRepo.scala:129]: updated_at needs to be updated

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala [hunk 13]
```diff
@@ -767,6 +767,11 @@ class UserCommander @Inject() (
     userRepo.save(currentUser.copy(primaryEmailId = Some(primaryEmail.id.get)))
     heimdalClient.setUserProperties(primaryEmail.userId, "$email" -> ContextStringData(primaryEmail.address))
   }
+
+  def getUserImageUrl(userId: Id[User], width: Int): Future[String] = {
+    val user = db.readOnly { implicit session => userRepo.get(userId) }
+    s3ImageStore.getPictureUrl(Some(width), user, user.pictureName.getOrElse("0"))
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:773]: Sorry :(

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/EmailOptOutController.scala (hunk 10 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/EmailOptOutController.scala:68]: We do, the domain is in the config, and the rest is the reverse router.

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/PimpedFuture.scala]: @andrewconner thanks, will remove the `toSeq`.
@yingjieMiao In my particular use case I am looking up the image url for several users. More generically, I have a exactly what you said, namely a `Seq[A]` and a function `A -> Future[B]`. Having a primitive that takes these two and produces a `Future[Map[A,B]]` would work just as well, but I felt like this is somewhat more generic and parallels the `Future.sequence` API in the standard lib.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 3 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:128]: Well, it is a `NonUserThread` :-)

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 5 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala]: thanks, fixing

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 6 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:136]: If we don't have an actual title, for now yeah. We'll have to play with it a little in QA. I have a feeling the design might need to be adapted slightly for sites with no title.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 9 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:168]: We decided not to initially because it could get very long very quickly with lots of participants and/or change when someone is added which might break threading. Also something to play around with in QA.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThread.scala (hunk 11 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThread.scala]: I was trying to give it behavior similar to a user which always has a `firstName` and a `lastName`. Not super happy with the usage either. What do you mean by "Always Presentation?"

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThreadRepo.scala (hunk 12 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/NonUserThreadRepo.scala:129]: good find, fixing.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 7 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:138]: Up until last night I wasn't entirely clear on how the templates should behave when these are missing, hence this. Working on those today (will be separate pull request).

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala (hunk 8 above)
Comment (stkem) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/NotificationCommander.scala:159]: There are three kinds of `MessageSender`, `User`, `NonUser` and `System`. `System` is being filtered out a few lines up.
